### PR TITLE
Added command to duplicate files and folders.

### PR DESCRIPTION
**Pull request by jbicker**
Fixes #GH-1737

### CODE REVIEW

File: packages/workspace/src/browser/workspace-duplicate-handler.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+
```
Comment (vince-fugnitto) [inline packages/workspace/src/browser/workspace-duplicate-handler.ts]: remove extra line

File: packages/workspace/src/browser/workspace-duplicate-handler.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+/********************************************************************************
+ * Copyright (C) 2017 TypeFox and others.
```
Comment (vince-fugnitto) [inline packages/workspace/src/browser/workspace-duplicate-handler.ts]: 2018

**Comment by vince-fugnitto** [comment]
I noticed a few strange behaviours while quickly testing. 

The command `duplicate` works but only displays the new 'copy' file once the browser is refreshed.
I also noticed that if commands are used on the new copied files (ex: `delete`), the browser must also be refreshed to view updates.

File: packages/workspace/src/browser/workspace-duplicate-handler.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    async execute(uris: URI[]): Promise<void> {
+        const distinctUris = URI.getDistinctParents(uris);
+        await Promise.all(distinctUris.map(uri => this.duplicate(uri)));
+    }
+
+    duplicate(uri: URI) {
+        try {
+            let name = uri.displayName;
+            const nameParts = name.split('.');
+            nameParts[0] = nameParts[0] + ' copy';
```
Comment (akosyakov) [inline packages/workspace/src/browser/workspace-duplicate-handler.ts]: What is there is already a file with such name? Please look at how other handlers deal with it, there should be a function to generate a unique name.

File: packages/workspace/src/browser/workspace-duplicate-handler.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            name = nameParts.join('.');
+            const target = new URI(uri.parent.path.toString()).resolve(name);
```
Comment (akosyakov) [inline packages/workspace/src/browser/workspace-duplicate-handler.ts]: `uri.parent.resovle(name)`?

File: packages/workspace/src/browser/workspace-duplicate-handler.ts (hunk 2 above)
Comment (jbicker) [reply packages/workspace/src/browser/workspace-duplicate-handler.ts]: OMG, my copy/paste blindness.
Thanks!

**Comment by jbicker** [comment]
Ok, refactored everything resp. reimplemented the way you suggested, @vince-fugnitto , @akosyakov.

File: packages/workspace/src/browser/workspace-commands.ts [hunk 5]
```diff
@@ -192,6 +196,25 @@ export class WorkspaceCommandContribution implements CommandContribution {
...
+        registry.registerCommand(WorkspaceCommands.FILE_DUPLICATE, this.newMultiUriAwareCommandHandler({
+            execute: async uris => {
+                await Promise.all(uris.map(async uri => {
+                    let parent = await this.getDirectory(uri);
+                    if (parent && parent.isDirectory) {
+                        parent = await this.getParent(uri);
+                    }
+                    if (parent) {
+                        const parentUri = new URI(parent.uri);
+                        const target = this.findVacantChildUri(parentUri, parent, uri.displayName, '-copy');
```
Comment (akosyakov) [inline packages/workspace/src/browser/workspace-commands.ts]: Should we really change the file extension?

File: packages/workspace/src/browser/workspace-commands.ts [hunk 6]
```diff
@@ -192,6 +196,25 @@ export class WorkspaceCommandContribution implements CommandContribution {
                 }
             })
         }));
+        registry.registerCommand(WorkspaceCommands.FILE_DUPLICATE, this.newMultiUriAwareCommandHandler({
+            execute: async uris => {
+                await Promise.all(uris.map(async uri => {
+                    let parent = await this.getDirectory(uri);
```
Comment (akosyakov) [inline packages/workspace/src/browser/workspace-commands.ts]: Should not be `getParent` be enough?
```ts
const parent = await this.getParent(uri);
if (parent) {
```

Comment (jbicker) [reply packages/workspace/src/browser/workspace-commands.ts]: Damn, you're right...

**Comment by jbicker** [comment]
Ok, thanks for your patience. Works well and names of copied files are looking pretty now.

File: packages/workspace/src/browser/workspace-commands.ts [hunk 7]
```diff
@@ -192,6 +196,30 @@ export class WorkspaceCommandContribution implements CommandContribution {
                 }
             })
         }));
+        registry.registerCommand(WorkspaceCommands.FILE_DUPLICATE, this.newMultiUriAwareCommandHandler({
+            execute: async uris => {
+                await Promise.all(uris.map(async uri => {
+                    const parent = await this.getParent(uri);
+                    if (parent) {
+                        const parentUri = new URI(parent.uri);
```
Comment (akosyakov) [inline packages/workspace/src/browser/workspace-commands.ts]: How about getting `name` and `ext` from `Path`?
```ts
const parentUri = new URI(parent.uri);
const name = uri.path.name + '_copy';
const ext = uri.path.ext;
const target = this.findVacantChildUri(parentUri, parent, name, ext);
```

Comment (jbicker) [reply packages/workspace/src/browser/workspace-commands.ts]: Too easy. ;-)

**Pull request merged**